lib/provider.go

```diff
@@ -91,44 +111,66 @@ func (p *Provider) Retrieve() (credentials.Value, error) {
 		window = time.Minute * 5
 	}
 
+	// TODO(nick): why are we using the source profile name and not the actual profile's name?
 	source := sourceProfile(p.profile, p.profiles)
-	session, name, err := p.sessions.Retrieve(source, p.SessionDuration)
-	p.defaultRoleSessionName = name
+	profileConf, ok := p.profiles[p.profile]
+	if !ok {
+		return credentials.Value{}, fmt.Errorf("missing profile named %s", p.profile)
+	}
+	key := sessioncache.OrigKey{
+		ProfileName: source,
+		ProfileConf: profileConf,
+		Duration:    p.SessionDuration,
+	}
+
+	var cachedSession *sessioncache.Session
+	cachedSession, err := p.sessions.Get(key)
 	if err != nil {
-		session, err = p.getSamlSessionCreds()
+		creds, err := p.getSamlSessionCreds()
```

Comment:
small nit: using the same variable name in this scope and the outer scope _might_ confuse a future engineer